### PR TITLE
Retry with rbf = -1 if tokens are dropped

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
# Description

Currently, when ragged buffer is set, we drop any tokens that exceed the buffer size. This can occur if the dataset is very imbalanced. 

This PR implements an optional flag to prevent token dropping. If it's set to true, then we will retry the entire train step if tokens are dropped with worse-case ragged buffer size to prevent token dropping. 

BUGS: b/555278394

# Tests

We saw retry successfully retries and has similar convergence to baseline when ragged buffer drops tokens (and diverges when retry is disabled), and when tokens were not dropped (rbf = 2) throughput is similar to baseline. Baseline experiments were executed on `cl/974137029` for direct comparison. All workloads evaluated on 256 TPU chips (DeepSeek-V3 671B MoE).

| Config | XID | Loss at Step 86 | Average Throughput over Last 60 Steps (Tokens/s/device) | Number of Retries (Logs) | CMD (gPaste Link) |
| :--- | :---: | :---: | :---: | :---: | :--- |
| **`p4head rbf = -1`** | [`286646082`](http://xids/286646082) | **`3.473`** *(ppl: 32.24)* | **`631.26`** *(mean step: 12.98s)* | 0 *(Dropless baseline)* | [gPaste: `deploy_p4head_7029_rbf_neg1`](https://paste.googleplex.com/6436702030266368) |
| **`p4head rbf = 2`** *(No Retry)* | [`286646089`](http://xids/286646089) | **`3.481`** *(ppl: 32.50)* | **`653.55`** *(mean step: 12.53s)* | 0 *(No retry)* | [gPaste: `deploy_p4head_7029_rbf_2`](https://paste.googleplex.com/6018910093180928) |
| **`pr rbf = -1`** *(No Retry)* | [`286718316`](http://xids/286718316) | **`3.457`** *(ppl: 31.72)* | **`632.81`** *(mean step: 12.95s)* | 0 *(Dropless baseline)* | [gPaste: `deploy_pr_6389_rbf_neg1`](https://paste.googleplex.com/5320189965631488) |
| **`pr rbf = 0.125`** *(No Retry)* | [`286646092`](http://xids/286646092) | **`5.271`** *(ppl: 194.57)* | **`949.74`** *(mean step: 8.63s)* | 0 *(Dropping without retry; diverged)* | [gPaste: `deploy_cl_rbf_0p125_no_retry`](https://paste.googleplex.com/6602725584928768) |
| **`pr rbf = 0.125 with retry`** | [`286621402`](http://xids/286621402) | **`3.477`** *(train ppl: 32.35)* | **`391.64`** *(mean step: 20.92s)* | **87 retries** *(All 87 steps retried; converged)* | [gPaste: `deploy_cl_rbf_0p125_retry`](https://paste.googleplex.com/4870707075481600) |
| **`pr rbf = 2 with retry`** | [`286718340`](http://xids/286718340) | **`3.468`** *(ppl: 32.07)* | **`653.84`** *(mean step: 12.53s)* | **0** *(0 overflows / 0 retries)* | [gPaste: `deploy_pr_6389_rbf_2_retry`](https://paste.googleplex.com/5304902767738880) |
| **`pr rbf = 2 without retry`** | [`286719268`](http://xids/286719268) | **`3.482`** *(ppl: 32.52)* | **`653.48`** *(mean step: 12.54s)* | **0** *(No retry comparison)* | [gPaste: `deploy_pr_6389_rbf_2_no_retry`](https://paste.googleplex.com/5579687863123968) |

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -1069,9 +1071,14 @@ def permute(
           local_num_experts,
           axis=0,
       )
+      local_overflow = (jnp.sum(local_group_size) > buffer_size).astype(jnp.int32)
       # Clamp local_group_size to buffer_size to ensure we don't exceed buffer
       # capacity by leveraging the helper _truncate_matrix.
       local_group_size = _truncate_matrix(local_group_size[:, None], buffer_size)[:, 0]
+      if num_expert_parallelism > 1:
+        has_overflow = jax.lax.psum(local_overflow, self._expert_parallelism_name) > 0
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: Don't we have to check all (FS)DP ranks as well? DP rank 0 might not overflow for any EP but DP rank 1 would overflow - these ranks will have different values for has_overflow. Can we just all reduce over every device?

File: tests/unit/maxtext_utils_test.py [hunk 2]
```diff
@@ -1088,6 +1088,15 @@ def test_donate_argnums_is_zero(self):
     )
     self.assertEqual(donate_argnums, 0)
 
+  def test_donate_argnums_is_zero_with_retry_when_tokens_dropped(self):
```
Comment (gobbleturk) [inline tests/unit/maxtext_utils_test.py]: Does flipping the value of retry_when_tokens_drop change the signature of train? I'm not sure whats being tested here that is specific to retry_when_tokens_dropped

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: good catch, updated

File: tests/unit/maxtext_utils_test.py (hunk 2 above)
Comment (Shuwen-Fang) [reply tests/unit/maxtext_utils_test.py]: It does with the updated version

File: src/maxtext/configs/base.yml [hunk 3]
```diff
@@ -234,6 +233,7 @@ ragged_buffer_factor: -1.0 # a factor to determine the size of the ragged buffer
 # When set to 1.0 this buffer if set to the size assuming perfectly balanced. If the routing dictates
 # a size larger than this then tokens are dropped.
 # In general if ragged_buffer_factor > 0, the ragged_buffer_size is balanced_size * ragged_buffer_factor.
+retry_when_tokens_dropped: false # retry a layer with worst-case ragged buffer size if tokens would otherwise be dropped.
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml]: Change comment to retry the entire train step -not just a single layer

File: src/maxtext/trainers/pre_train/train.py [hunk 4]
```diff
@@ -885,7 +890,26 @@ def training_loop_iteration(
       with jax.set_mesh(mesh), nn_partitioning.axis_rules(logical_axis_rules_for_train):
         if shard_optimizer_over_data and isinstance(model, nn.Module):
           state = sharding.maybe_shard_with_name(state, state_mesh_shardings, shard_mode)
-        state, metrics = p_train_step(state, example_batch, *step_rng_args)
+        if config.retry_when_tokens_dropped and p_train_step_dropless is not None:
+          candidate_state, metrics = p_train_step(state, example_batch, *step_rng_args)
+          if bool(metrics.get("has_moe_overflow")):
```
Comment (gobbleturk) [inline src/maxtext/trainers/pre_train/train.py:926]: This flow requires the metrics be ready immediately following the jitted function dispatch (blocking on it finishing), which may prevent the overlap of data loading

Have you tested this in in our target setting with a real dataset?

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/trainers/pre_train/train.py (hunk 4 above)
Comment (Shuwen-Fang) [reply src/maxtext/trainers/pre_train/train.py:926]: Discussed offline; tested with no perf diff

File: src/maxtext/configs/base.yml (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/configs/base.yml]: Curious why the retry didn't happen at moe.py scope?

Comment (RissyRan) [reply src/maxtext/configs/base.yml]: Could you help add this flag into https://github.com/AI-Hypercomputer/maxtext/blob/main/docs/reference/core_concepts/moe_configuration.md?

File: src/maxtext/configs/types.py [hunk 5]
```diff
@@ -882,6 +873,10 @@ class MoEGeneral(BaseModel):
   num_experts: PositiveInt = Field(1, description="The total number of experts in each MoE layer.")
   num_experts_per_tok: PositiveInt = Field(1, description="The number of experts to route each token to.")
   capacity_factor: float = Field(-1.0, description="Expert capacity factor. If < 0, no token dropping.")
+  retry_when_tokens_dropped: bool = Field(
```
Comment (RissyRan) [inline src/maxtext/configs/types.py]: Nit: let's put under ragged_buffer_size?

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -3291,6 +3252,20 @@ def _validate_check_vma_is_supported(self):
           f"Found other ICI axes enabled: {active}."
       )
 
+  def validate_retry_when_tokens_dropped(self):
+    """Validates prerequisites for the per-layer dropless fallback."""
+    if self.retry_when_tokens_dropped:
+      if self.num_experts <= 1:
+        raise ValueError("retry_when_tokens_dropped=True requires num_experts > 1.")
+      if self.ragged_buffer_factor <= 0:
```
Comment (RissyRan) [inline src/maxtext/configs/types.py:3354]: Nit: == -1 for default

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -481,6 +434,7 @@ def __init__(
       dtype: ctypes.DType = jnp.float32,
       quant: Optional[quantizations.AqtQuantization] = None,
       is_hash_routing: bool = False,
+      force_dropless_buffer: bool = False,
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py]: Nit: for bool maybe force_dropless?

File: src/maxtext/layers/moe.py [hunk 8]
```diff
@@ -1105,7 +1062,7 @@ def permute(
 
     num_tokens = bsz_times_seq_len * self.num_experts_per_tok
     use_truncated_buffer = use_ragged_in_permute and buffer_size is not None and buffer_size < num_tokens
-
+    has_overflow = jnp.bool_(False)
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py]: Nit: What's the benefit using jnp.bool_?

File: src/maxtext/configs/base.yml (hunk 3 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/base.yml]: Because if we retry in moe.py (within jax.jit), then we observe extra overhead at run time: https://b.corp.google.com/issues/555278394#comment4

**Review by RissyRan — APPROVED** [review]
Thanks for the change! Let's follow up if any other perf issue comes.

File: src/maxtext/layers/moe.py (hunk 8 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/layers/moe.py (hunk 7 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/configs/types.py (hunk 6 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/types.py:3354]: updated

File: src/maxtext/configs/types.py (hunk 5 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/types.py]: updated
